### PR TITLE
feat: capture emoji reactions as a reinforcement signal (#27438)

### LINKED ISSUE #27438: Feature: Emoji Reaction Reinforcement — Learning from 👍 / ❤️ / 👎 on messaging platforms
## Problem

Hermes lives on Telegram, Discord, Slack, and other messaging platforms where users naturally react to messages with emoji. Right now those reactions are just… decor. Thumbs up, heart, thumbs down, poo — they’re rich, low-friction feedback signals that go completely unused.

Every time a user reacts to a Hermes message, they’re implicitly saying "good job," "I love this," "that was crap," or "wtf was that." That’s free training data, and it requires zero extra effort from the user — they’re already doing it.

## Proposal

Capture emoji reactions on supported messaging channels and map them into a lightweight reinforcement signal that Hermes can act on.

### Emoji → Signal Mapping

| Reaction | Signal | Weight |
|----------|--------|--------|
| ❤️ heart | Overwhelmingly positive | +2.0 |
| 👍 thumbs up | Positive | +1.0 |
| 😂 laugh | Positive/entertaining | +0.8 |
| 🙌 raised hands | Positive/celebratory | +1.0 |
| 👎 thumbs down | Negative | -1.0 |
| 💩 poo | Strongly negative | -2.0 |
| 😢 cry | Negative/disappointing | -1.5 |
| 😡 angry | Strongly negative | -2.0 |

Configurable per user — some people just react with everything.

### What "Learning" Means Here

This is NOT model fine-tuning (that’s #498’s territory). This is about lightweight, in-process reinforcement:

1. **Memory weighting** — Messages that receive positive reactions get higher retrieval priority in future context lookups. Messages that get negative reactions get deprioritized or flagged for review.

2. **Skill confidence scoring** — If a skill is active when a positive reaction comes in, bump its internal confidence/usage score. Negative reaction → dock it. Over time, Hermes learns which skills actually work vs. which ones look good on paper.

3. **Response style tuning** — Track which tone/length/formats get positive reactions per user and per channel, and bias future responses accordingly (without overriding explicit instructions).

4. **Preference extraction** — "You reacted with ❤️ to my movie recommendations but 👎 to my recipe suggestions" → that’s a user preference signal worth persisting.

### Implementation Considerations

- **Scope:** Reactions are channel-scoped and user-scoped. A reaction in a group chat shouldn’t necessarily affect the same user’s private chat behavior (or maybe it should — make it configurable).
- **Noise handling:** Some users react to everything. Include a decay factor and require a minimum signal threshold before acting.
- **Privacy:** This is per-user, per-instance. Not telemetry. Not shared.
- **Transparency:** Hermes should occasionally acknowledge that it’s learning from reactions ("Noted, I’ll stop suggesting Python when you clearly prefer Go"). Not every time — that’d be annoying — but occasionally.

### Relationship to Existing Work

- **#498 (Conversational RL Personalization):** This is the heavy-duty approach — actual model weight updates. The emoji reaction system could *feed into* #498 as a reward signal source, but it’s independently valuable as a lightweight mechanism that works today.
- **#337 (Skill Evolution):** Reaction feedback could inform which skills get auto-improved vs. deprecated.

### Why This Matters

The best feedback systems are the ones users don’t have to think about. Typing "that was helpful" is work. Tapping ❤️ is muscle memory. If Hermes is going to be "the agent that grows with you," it should listen to the signals users are already sending.

### PR DESCRIPTION
**Pull request by xxxigm**
## What does this PR do?

Implements the v1 capture layer for [#27438 — Emoji Reaction Reinforcement](https://github.com/NousResearch/hermes-agent/issues/27438): Hermes now turns 👍 / ❤️ / 👎 / 💩 / … reactions on its own messages into a lightweight, durable feedback signal instead of letting them disappear.

The PR is deliberately scoped to the **capture + storage foundation + one platform** so it stays reviewable:

1. **`gateway/reactions.py`** — pure module with `ReactionSignal`, `ReactionPolarity`, `ReactionEvent`, the `DEFAULT_REACTION_WEIGHTS` table lifted straight from the issue, variation-selector-aware `resolve_emoji_signal()`, and a `ReactionConfig.from_env()` loader. No side effects, import-safe.
2. **`gateway/reaction_store.py`** — thread-safe SQLite-backed `ReactionStore` at `$HERMES_HOME/reactions.db` with the standard WAL-with-DELETE fallback (matches `state.db` / `kanban.db`). Exposes `record()`, `aggregate_for_message()`, `recent_for_user()`, and opt-in `prune_older_than()` for retention.
3. **`BasePlatformAdapter.handle_reaction()`** — new default that honours the master `HERMES_REACTION_SIGNALS_ENABLED` flag, persists via the store, and fires a new `on_reaction` lifecycle hook so future consumers (memory weighting, skill confidence) can subscribe without touching individual adapters.
4. **Telegram wiring** — register PTB's `MessageReactionHandler` (PTB 22.6 is already pinned) and translate `MessageReactionUpdated` into `ReactionEvent`s via set-diff (so swapping 👍→❤️ emits one removed + one added event, not a rewrite).
5. **Config** — new top-level `reaction_signals:` block in `DEFAULT_CONFIG` + YAML→env bridge in `gateway/config.py`. Strictly **separate** from the per-platform `*.reactions` toggles (which control the outgoing 👀/✅/❌ lifecycle markers) so neither feature can clobber the other.
6. **Tests** — 95 new tests across three new files covering the pure module, the SQLite store (schema, indexes, aggregation, removal subtraction, retention), and the Telegram handler (set-diff, anonymous-reaction drop, master-flag enforcement, end-to-end PTB-update → DB row).
7. **Docs** — new user-guide page, env var reference section, and Telegram-guide cross-reference.

The actual *consumers* of the signal (memory weighting, skill confidence scoring, response-style tuning, preference extraction) are deliberately deferred to follow-up PRs — they will read events from `ReactionStore` without further changes to the capture layer. Discord (`on_raw_reaction_add`) and Slack (`reaction_added`) adapters can layer on the same `BasePlatformAdapter.handle_reaction()` hook in their own PRs.

The feature is **strictly opt-in**: `HERMES_REACTION_SIGNALS_ENABLED` defaults to `false`, so existing deployments see no behavioural change.

## Related Issue

Closes #27438.

Independently valuable from #498 (Conversational RL Personalization — heavy-duty model fine-tuning) but designed so the captured events could later feed #498 as a reward source. Captured events could also inform #337 (Skill Evolution) to decide which skills get auto-improved.

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 🔒 Security fix
- [x] 📝 Documentation update
- [x] ✅ Tests (adding or improving test coverage)
- [ ] ♻️ Refactor (no behavior change)
- [ ] 🎯 New skill (bundled or hub)

## Changes Made

**Core (new files):**

- `gateway/reactions.py` (+253) — `ReactionSignal`, `ReactionPolarity`, `ReactionEvent`, `DEFAULT_REACTION_WEIGHTS`, `resolve_emoji_signal()`, `ReactionConfig.from_env()`, `build_reaction_event()`.
- `gateway/reaction_store.py` (+292) — `ReactionStore` (WAL fallback, schema v1, three indexes), `get_reaction_store()` singleton, `default_reactions_db_path()` (HERMES_HOME-aware, resolved per call), `reset_reaction_store_for_tests()`.

**Platform layer:**

- `gateway/platforms/base.py` — new `handle_reaction()` default + `on_reaction()` extension point; safe import of `ReactionEvent` for the signature.
- `gateway/platforms/telegram.py` — lazy import of `MessageReactionHandler` (PTB 20.7+), registration gated on `HERMES_REACTION_SIGNALS_ENABLED`, `_handle_message_reaction()` callback with old/new set-diff and chat-type propagation. `Update.ALL_TYPES` was already used by `start_polling()` so no `allowed_updates` change is required.

**Config:**

- `hermes_cli/config.py` — new `reaction_signals:` block in `DEFAULT_CONFIG`.
- `gateway/config.py` — YAML→env bridge for `enabled`, `min_signal_threshold`, `decay_days`, `include_unknown`.

**Tests (new files, 95 new test cases):**

- `tests/gateway/test_reactions.py` — 50 cases covering the canonical emoji table, polarity-aware `signed_weight()`, VS16 normalisation, env-var parsing, override merging.
- `tests/gateway/test_reaction_store.py` — 29 cases covering schema bootstrap, idempotent re-init, record/aggregate/recent/prune contracts, removal-subtraction, since-filter, profile-scoped path resolution, singleton lifecycle.
- `tests/gateway/test_telegram_reaction_signals.py` — 16 cases covering master-flag enforcement on both ends, add/remove/swap diff, anonymous-reaction drop, malformed-update tolerance, unknown-emoji include/exclude policy, and an end-to-end PTB-update → SQLite-row integration test.

**Docs:**

- `website/docs/user-guide/reaction-reinforcement.md` — new page with defaults table, enable instructions, Telegram group-admin caveat, inspection snippets, retention pruning, troubleshooting.
- `website/docs/reference/environment-variables.md` — new "Emoji Reaction Reinforcement" section, explicitly contrasted with the existing `*_REACTIONS` lifecycle vars.
- `website/docs/user-guide/messaging/telegram.md` — cross-reference subsection in the existing "Message Reactions" block.

## How to Test

1. Check out this branch and ensure the venv is set up:
   ```
   python3 -m venv .venv && source .venv/bin/activate && pip install -e ".[all,dev]"
   ```
2. Run the new test files in isolation:
   ```
   scripts/run_tests.sh tests/gateway/test_reactions.py tests/gateway/test_reaction_store.py tests/gateway/test_telegram_reaction_signals.py -v
   ```
   Expected: **95 passed**.
3. Run the new tests **plus** the existing Telegram-reactions suite to confirm no overlap or regression with the lifecycle (👀/✅/❌) feature:
   ```
   scripts/run_tests.sh tests/gateway/test_reactions.py tests/gateway/test_reaction_store.py tests/gateway/test_telegram_reaction_signals.py tests/gateway/test_telegram_reactions.py
   ```
   Expected: **116 passed**.
4. Manual smoke test on a Telegram bot:
   ```
   export HERMES_REACTION_SIGNALS_ENABLED=true
   hermes gateway restart
   # In Telegram, react to a bot message with 👍, ❤️, 👎.
   sqlite3 ~/.hermes/reactions.db \
     "SELECT ts, platform, emoji, label, weight, added \
        FROM reaction_events ORDER BY ts DESC LIMIT 10;"
   ```
   Expected: one row per reaction tap with the correct emoji / label / weight.
5. Confirm the feature is a no-op when disabled (the default):
   ```
   unset HERMES_REACTION_SIGNALS_ENABLED
   # Restart gateway, react to a bot message.
   ls ~/.hermes/reactions.db   # File should NOT be created.
   ```

### Code

- [x] I've read the [Contributing Guide](https://github.com/NousResearch/hermes-agent/blob/main/CONTRIBUTING.md)
- [x] My commit messages follow [Conventional Commits](https://www.conventionalcommits.org/) (`feat(reactions):`, `feat(platforms):`, `docs(reactions):`)
- [x] I searched for [existing PRs](https://github.com/NousResearch/hermes-agent/pulls) to make sure this isn't a duplicate
- [x] My PR contains **only** changes related to this feature (no unrelated commits)
- [x] I've run `scripts/run_tests.sh` and all tests pass (116/116 on the relevant suites)
- [x] I've added tests for my changes (95 new test cases)
- [x] I've tested on my platform: macOS 15.2 (Darwin 24.6.0), Python 3.12

### Documentation & Housekeeping

- [x] I've updated relevant documentation — new user-guide page (`reaction-reinforcement.md`), env-vars reference section, and Telegram-guide cross-reference
- [x] I've updated `cli-config.yaml.example` if I added/changed config keys — N/A (top-level `reaction_signals:` block lives in `DEFAULT_CONFIG`; the example file does not enumerate every default)
- [x] I've updated `CONTRIBUTING.md` or `AGENTS.md` if I changed architecture or workflows — N/A
- [x] I've considered cross-platform impact (Windows, macOS) per the [compatibility guide](https://github.com/NousResearch/hermes-agent/blob/main/CONTRIBUTING.md#cross-platform-compatibility) — `ReactionStore` reuses the existing WAL-with-DELETE fallback so NFS / SMB / WSL1 behave identically to `state.db`
- [x] I've updated tool descriptions/schemas if I changed tool behavior — N/A (no tool surface change in this PR)

## Screenshots / Logs

```
$ scripts/run_tests.sh tests/gateway/test_reactions.py tests/gateway/test_reaction_store.py tests/gateway/test_telegram_reaction_signals.py tests/gateway/test_telegram_reactions.py
4 workers [116 items]
============================= 116 passed in 1.31s ==============================
```

Example aggregate-for-message output after a couple of taps:

```
>>> store.aggregate_for_message(platform="telegram", channel_id="42",
...                             target_message_id="123")
{'platform': 'telegram', 'channel_id': '42', 'target_message_id': '123',
 'net_weight': 2.0, 'positive': 1, 'negative': 0, 'neutral': 0,
 'unique_users': 1, 'sample_count': 1}
```

### CODE REVIEW

**Comment by teknium1** [comment]
Thanks for building a thorough capture/storage foundation; current main still does not have this feature. A few issues should be addressed before maintainers salvage it.

### Problems

- `gateway/platforms/telegram.py:2662-2746` on PR head `1cbafb9d4f9a6eecec4344976582451c452cb362` records any `message_reaction` update with a user/chat/message ID. It does not verify that `target_message_id` is a Hermes-authored outbound message, even though the PR scope is reactions on Hermes messages.
- `gateway/reaction_store.py:223-234` subtracts removed reactions from `net_weight`, but `positive`/`negative`/`neutral` counts and `unique_users` only increment on added rows. Add→remove of 👎 leaves `negative=1` and `unique_users=1` even though the net state is cleared.
- `website/docs/user-guide/reaction-reinforcement.md:63-65` and `website/docs/reference/environment-variables.md:515-518` add user-facing `HERMES_REACTION_*` settings for behavioral config. `AGENTS.md:104-108` says new non-secret behavior settings should be `config.yaml`-first, not new `HERMES_*` env vars.

### Suggested changes

- Correlate reaction updates to a bounded set of Hermes-authored outbound message IDs and drop unknown/non-Hermes targets.
- Make aggregation counts match the intended removal semantics, with an add-then-remove test.
- Keep `reaction_signals:` as the user-facing config surface and remove/avoid new non-secret `HERMES_*` docs unless maintainers explicitly approve them.

Automated hermes-sweeper review.

**Comment by xxxigm** [comment]
Thanks @teknium1 — all three points addressed, and I merged current `main` in so the branch is no longer conflicting.

1. **Reactions are now scoped to Hermes-authored messages.** `BasePlatformAdapter` keeps a bounded LRU of `(channel_id, message_id)` keys it sent (`record_authored_message` / `is_authored_message`), and `handle_reaction` drops any reaction whose target isn't in that set — a reaction on a user's / third-party message is no longer recorded. Telegram's `send()` records every outbound message id.

2. **Removal aggregation fixed.** `aggregate_for_message` now reconstructs the net per-`(user, emoji)` state in time order, so a removal cancels its matching add across the **whole** aggregate — `net_weight`, the polarity counts, AND `unique_users`. Previously an add-then-remove of 👎 left `negative=1` / `unique_users=1`. Added regression tests for add→remove (counts zero out), 👎 add→remove, and a 👍→❤️ swap.

3. **config.yaml-first.** `ReactionConfig.from_env()` now reads the `reaction_signals:` block from `config.yaml` as the authoritative surface; the `HERMES_REACTION_*` env vars are retained only as an internal override. Docs (`reaction-reinforcement.md`, telegram, env reference) point at `config.yaml`, and `reaction_signals` is registered in `_KNOWN_ROOT_KEYS`. Added tests for config-enable, env-overrides-config, missing-config defaults, and malformed-config-doesn't-wedge.

**Review by teknium1 — COMMENTED** [review]
Thanks for the thorough capture/store design and for addressing the earlier ownership, aggregation, and config-surface feedback. The underlying inbound-reaction gap still exists on current main, but this needs a current-platform salvage rather than a direct merge.

### Problems
- `gateway/platforms/telegram.py:2415` records authored IDs only in the legacy send loop. Current main can return through the rich-message path at `plugins/platforms/telegram/adapter.py:3592-3607`, so reactions to rich replies would not pass the authored-message check.
- `gateway/reactions.py:247-248` exposes `min_signal_threshold` and `decay_days`, but the PR has no runtime threshold consumer and retention only happens through a manual `prune_older_than()` call.
- Telegram moved from `gateway/platforms/telegram.py` to `plugins/platforms/telegram/adapter.py` in `5600105478ffde29d7566b45421b100eaa29c4ef`; the PR is currently dirty/unmergeable.

### Suggested changes
- Port the integration to the bundled Telegram plugin and cover current rich/stream delivery paths when recording authored IDs.
- Defer the unused config knobs, or add the concrete consumer and lifecycle that makes them effective.

Automated hermes-sweeper review.

File: gateway/platforms/telegram.py [hunk 1]
```diff
@@ -2371,6 +2410,9 @@ async def send(
                                 continue
                         raise
                 message_ids.append(str(msg.message_id))
+                # Remember this outbound message so reaction signals (#27438)
+                # can be correlated to Hermes-authored messages only.
+                self.record_authored_message(chat_id, msg.message_id)
```
Comment (teknium1) [inline gateway/platforms/telegram.py:2415]: This only marks messages sent through the legacy chunk loop. Current main's Telegram adapter can return successfully from `sendRichMessage` before reaching that loop (`plugins/platforms/telegram/adapter.py:3592-3607`), so the ownership check will drop reactions to rich replies. Centralize this tracking across every successful outbound delivery path.

File: gateway/reactions.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+    Loaded from ``config.yaml``'s ``reaction_signals:`` block (the
+    authoritative user-facing surface) in :func:`from_env`, with the
+    ``HERMES_REACTION_*`` env vars retained only as an internal override.
+    Defaults preserve today's behaviour (no-op) so this feature is strictly
+    opt-in.
+    """
+
+    enabled: bool = False
+    min_signal_threshold: float = 0.5
```
Comment (teknium1) [inline gateway/reactions.py:247]: `min_signal_threshold` is exposed as configuration but no capture or aggregation path consumes it in this PR. Please defer this knob until its first consumer ships, or implement that consumer and test its behavior.
